File: src/templates/c/lib_utils.c
// ═══════════════════════════════════════════════════════════════
// LIBRERÍA: Utilidades Básicas (C)
// Nivel: Básico - Componente independiente
// ═══════════════════════════════════════════════════════════════
// Funciones utilitarias de uso común.
// Hereda a otros nodos para añadir funcionalidades básicas.

#include <stdio.h>
#include <stdlib.h>

// ═══════════════════════════════════════════════════════════════
// Macros de utilidad
// ═══════════════════════════════════════════════════════════════
#define MAX(a, b) ((a) > (b) ? (a) : (b))
#define MIN(a, b) ((a) < (b) ? (a) : (b))
#define ABS(x) ((x) < 0 ? -(x) : (x))
#define SWAP(a, b, type) { type temp = a; a = b; b = temp; }

// ═══════════════════════════════════════════════════════════════
// Funciones de impresión formateada
// ═══════════════════════════════════════════════════════════════
void print_separator(const char* title) {
    printf("\n══════════════════════════════════════\n");
    printf("  %s\n", title);
    printf("══════════════════════════════════════\n");
}

void print_int(const char* label, int value) {
    printf("%s: %d\n", label, value);
}

void print_float(const char* label, float value) {
    printf("%s: %.2f\n", label, value);
}

// ═══════════════════════════════════════════════════════════════
// Ejemplo de uso
// ═══════════════════════════════════════════════════════════════
int main() {
    print_separator("Librería de Utilidades");
    
    int a = 10, b = 5;
    print_int("MAX(10, 5)", MAX(a, b));
    print_int("MIN(10, 5)", MIN(a, b));
    print_int("ABS(-7)", ABS(-7));
    
    SWAP(a, b, int);
    print_int("Después de SWAP, a", a);
    print_int("Después de SWAP, b", b);
    
    return 0;
}

File: src/templates/java/lib_utils.java
// ═══════════════════════════════════════════════════════════════════════════
// Java 25 - Librería de Utilidades
// ═══════════════════════════════════════════════════════════════════════════
// Esta librería proporciona funciones de utilidad reutilizables
// que pueden ser heredadas por otros nodos usando ch("lib_utils")

import java.util.*;
import java.util.function.*;
import java.util.stream.Collectors;

public class Utils {
    // Validación de entrada
    public static <T> T requireNonNull(T obj, String mensaje) {
        if (obj == null) {
            throw new IllegalArgumentException(mensaje);
        }
        return obj;
    }
    
    // Verificar rango
    public static void requireRange(int value, int min, int max) {
        if (value < min || value > max) {
            throw new IllegalArgumentException(
                String.format("Valor %d fuera del rango [%d, %d]", value, min, max)
            );
        }
    }
    
    // Retry con reintentos
    public static <T> T retry(Supplier<T> operation, int maxAttempts) {
        Exception lastException = null;
        
        for (int i = 0; i < maxAttempts; i++) {
            try {
                return operation.get();
            } catch (Exception e) {
                lastException = e;
                if (i < maxAttempts - 1) {
                    try {
                        Thread.sleep(100 * (i + 1)); // Backoff exponencial
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new RuntimeException(ie);
                    }
                }
            }
        }
        
        throw new RuntimeException("Operación falló después de " + maxAttempts + " intentos", lastException);
    }
    
    // Medir tiempo de ejecución
    public static <T> T measureTime(String operationName, Supplier<T> operation) {
        long start = System.nanoTime();
        try {
            return operation.get();
        } finally {
            long duration = System.nanoTime() - start;
            System.out.println(operationName + " tomó: " + (duration / 1_000_000.0) + " ms");
        }
    }
    
    // Transformar lista
    public static <T, R> List<R> map(List<T> lista, Function<T, R> mapper) {
        return lista.stream().map(mapper).toList();
    }
    
    // Filtrar lista
    public static <T> List<T> filter(List<T> lista, Predicate<T> predicate) {
        return lista.stream().filter(predicate).toList();
    }
    
    // Agrupar por criterio
    public static <T, K> Map<K, List<T>> groupBy(List<T> lista, Function<T, K> classifier) {
        return lista.stream().collect(
            java.util.stream.Collectors.groupingBy(classifier)
        );
    }
    
    // Ejemplo de uso
    public static void main(String[] args) {
        List<Integer> numeros = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
        
        // Filtrar pares
        List<Integer> pares = filter(numeros, n -> n % 2 == 0);
        System.out.println("Pares: " + pares);
        
        // Mapear a cuadrados
        List<Integer> cuadrados = map(numeros, n -> n * n);
        System.out.println("Cuadrados: " + cuadrados);
        
        // Medir tiempo
        measureTime("Operación costosa", () -> {
            try { Thread.sleep(100); } catch (InterruptedException e) {}
            return "Completado";
        });
    }
}

